ui/src/frontend/hsluv_cache.ts

```diff
@@ -37,3 +37,21 @@ export function cachedHsluvToHex(
     hue: number, saturation: number, lightness: number): string {
   return cache.get(hue, saturation, lightness);
 }
+
+const textColorCache = new Map<string, string>();
```

Comment:
nit: this collection isn't self documenting, as it uses primitive types. Additionally, the variable name doesn't really help other than reveal text colors are involved somehow. So, a comment is really called for. FWIW, here's what I would write:
```
// Key is background color in '#rrggbb' form. Value is suitable contrasting text color, in same form.
```